blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BluefloodCounterRollupTest.java

```diff
@@ -1,93 +1,758 @@
+        //          | <-- one time unit --> |                       |
+        //    ------+-----------------------+-----------------------+-----
+        //          1                       1                       1
+        //          2                       2                       2
+        //          3                       3                       3
+        //          4                       5                       6
+        //          L                       L                       L
+        //          |                                               |
+        //          | <-------------- two time units -------------> |
+        //                             (3+4+5)/2=6
+        //
+        //
+        // Samples As Intervals:
+        //                                                           _______________________
+        //                                   _______________________|          5            |
+        //           _______________________|          4            |                       |
+        //          |          3            |                       |                       |
+        //          |                       |                       |                       |
+        //          |                       |                       |                       |
+        //    ------+-----------------------+-----------------------+-----------------------+-----
+        //          1                       1                       1                       1
+        //          2                       2                       2                       2
+        //          3                       3                       3                       3
+        //          4                       5                       6                       7
+        //          L                       L                       L                       L
+        //          |                                                                       |
+        //          | <------------------------- three time units ------------------------> |
+        //                                         (3+4+5)/3=4
+        //
+
+        // given
+        Points<SimpleNumber> input = new Points<SimpleNumber>();
+        input.add(new Points.Point<SimpleNumber>(1234L, new SimpleNumber(3L)));
+        input.add(new Points.Point<SimpleNumber>(1235L, new SimpleNumber(4L)));
+        input.add(new Points.Point<SimpleNumber>(1236L, new SimpleNumber(5L)));
+        // when
+        BluefloodCounterRollup rollup = BluefloodCounterRollup.buildRollupFromRawSamples(input);
+        // then
+        Number count = rollup.getCount();
+        assertNotNull(count);
+        assertThat(count, is(CoreMatchers.<Number>instanceOf(Long.class)));
+        assertEquals(12L, count.longValue());
+        assertEquals(6.0d, rollup.getRate(), 0.00001d); // (3+4+5)/(3-1)=6
+        assertEquals(3, rollup.getSampleCount());
```

Comment:
Hi @izrik, you're right in that this is a tricky problem.  Personally, for rate, I think it's should measure values over time and average on number of points, not interval.

e.g.
if you start walking at 1am and travel 3 miles the first hour, 4 miles the 2nd hour, and 5 miles the 3rd hour, by 4am you traveled a total of 12 miles.  Your "rate" (or average velocity) per hour would be the total of all the miles walked: 12 miles (3 + 4 + 5), in the interval of 3 hours (4 - 1), so the rate would = 12 / 3 = 4.  

But the current code is actually treating rate as interva, because the three data points are collected at the 2am, 3am. and 4am mark (4 - 2 = 2).

For rate, we have to assume that the first data point is achieved by an interval occurring a second before the first point, which is the same as the number of data points we received at the regular interval (i.e. input.getPoints().size()).

So the code should look like this instead (quoted here with added comments by me):

```
public static BluefloodCounterRollup buildRollupFromRawSamples(Points<SimpleNumber> input) throws IOException {
        long minTime = Long.MAX_VALUE;
        long maxTime = Long.MIN_VALUE;
        BluefloodCounterRollup rollup = new BluefloodCounterRollup();
        Number count = 0L;
        for (Points.Point<SimpleNumber> point : input.getPoints().values()) {
            // count = sum of point values (3 + 4 + 5 = 12)
            count = sum(count, point.getData().getValue());

            // minTime = 1234
            minTime = Math.min(minTime, point.getTimestamp());

            // maxTime = 1236
            maxTime = Math.max(maxTime, point.getTimestamp());
        }

        // the problem is here.  Instead of using numSeconds of maxTime minus minTime (1236 - 1234 = 2)
        // it should actually be dividing by number of points from input.getPoints.size();
        // XXX double numSeconds = (double)(maxTime - minTime);
        // XXX double rate = count.doubleValue() / numSeconds;
        double rate = count.doubleValue() / input.getPoints().size();
        return rollup.withCount(count).withRate(rate).withSampleCount(input.getPoints().size());
    }
```


Can you step through